File: leetcode/letterCombinationsofaPhoneNumber.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <stack>
#include <queue>
#include "../utils.h"

using namespace std;

// Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent.

// A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.

// Example:

// Input: "23"
// Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
// Note:

// Although the above answer is in lexicographical order, your answer could be in any order you want.

class Solution {
public:
    void backtracking(unordered_map<string, string> &phone, vector<string> &output, string combination, string next_digits) {
        if (next_digits == "")
        {
            output.push_back(combination);
        }
        else
        {
            auto cur_digit = next_digits.substr(0,1);
            if (phone.find(cur_digit) != phone.end()) {
                auto chars = phone[cur_digit];
                for (auto &c : chars) {
                    backtracking(phone, output, combination + c, next_digits.substr(1));
                }
            }
        }
    }

    vector<string> letterCombinations(string digits) {
        auto phone = unordered_map<string, string>();
        phone["2"] = "abc";
        phone["3"] = "def";
        phone["4"] = "ghi";
        phone["5"] = "jkl";
        phone["6"] = "mno";
        phone["7"] = "pqrs";
        phone["8"] = "tuv";
        phone["9"] = "wxyz";
        
        vector<string> ans;
        if (digits != "")
            backtracking(phone, ans, "", digits);
        return ans;
    }
};

int main()
{
    assert(Solution().letterCombinations("23")
        == (vector<string> {"ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"}));
    return 0;
}

File: leetcode/leetcode_py/letter-combinations-of-a-phone-number.py
'''
Soltuion1. Backtracking
time: o(4^N) each digit, the algorithm explores all possible combinations
space: o(4^N) depth of the recursion
'''

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        m = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}

        res = []
        def backtracking(i, s):
            if i >= len(digits):
                if s:
                    res.append(s)
                return
            for c in m[digits[i]]:
                backtracking(i+1, s + c)

        backtracking(0, "")
        return res
        